examples/computer_vision/efficientdet_pytorch/model_def.py

```diff
@@ -0,0 +1,394 @@
+
+    def train_batch(self, batch: TorchData, epoch_idx: int, batch_idx: int):
+        
+        if epoch_idx != self.cur_epoch and self.lr_scheduler is not None:
+            self.cur_epoch = epoch_idx
+            self.num_updates = epoch_idx * self.data_length
+            
+            self.lr_scheduler.step(self.cur_epoch)
+            lrl2 = [param_group['lr'] for param_group in self.optimizer.param_groups]
+            print (self.cur_epoch, 'new lr: ', lrl2, batch_idx)
+
+        input, target = batch
+
+        if self.args.prefetcher:
+            input = input.float().sub_(self.train_mean).div_(self.train_std)
+            if self.train_random_erasing is not None:
+                input = self.train_random_erasing(input, target)
+
+        if self.args.channels_last:
+            input = input.contiguous(memory_format=torch.channels_last)
+
+        output = self.model(input, target)
+        loss = output['loss']
+
+        self.context.backward(loss)
+        self.context.step_optimizer(self.optimizer, self.clip_grads)
+        
+        if self.model_ema is not None:
+            self.model_ema.update(self.model)
+
+        self.num_updates += 1
+        if self.lr_scheduler is not None:
+            self.lr_scheduler.step_update(num_updates=self.num_updates) 
+            lrl = [param_group['lr'] for param_group in self.optimizer.param_groups]
+
+        return {"loss": loss}
+
+    def evaluate_batch(self, batch: TorchData):
+
+        input, target = batch
+        if self.args.prefetcher:
+            input = input.float().sub_(self.val_mean).div_(self.val_std)
+
+        if self.val_random_erasing is not None:
+            input = self.val_random_erasing(input, target)
+
+        input = self.context.to_device(input)
+        target = self.context.to_device(target)
+
+        output = self.model_ema.ema(input, target)
+        loss = output['loss']
+
+        reduced_loss = loss.data.item()
+
+        if reduced_loss is np.nan or math.isnan(reduced_loss):
+            # The original implementation is sensitive to the hyperparameters and configurations. 
+            # This is used to help debug.
+            for name, p in self.model.named_parameters():
+                print('Nan occurred: ',name,'norm: ', p.grad.norm().item(), 'sum: ', p.grad.sum().item(), 'max: ', p.grad.min().item(), 'min: ', p.grad.min().item())
+            reduced_loss = 0
+
+        if self.evaluator is not None:
+            self.evaluator.add_predictions(output['detections'], target)
+
+        vals = (self.evaluator.img_indices, self.evaluator.predictions)
+        self.val_reducer.update(vals)
+        self.evaluator.reset()
+
+        return {'val_loss': reduced_loss}   
+
+    def create_evaluator(self,name, dataset, pred_yxyx=False):
+        if 'coco' in name:
+            return CocoEvaluator(dataset, pred_yxyx=pred_yxyx)
+            
+    def validation_reducer(self, values):
```

Comment:
Can we use custom reduce metrics here or should the implementation be left as this?

Reply:
Offline disc: this is using the simple API